pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go

```diff
@@ -497,6 +497,32 @@ func init() {
 		},
 	)
 
+	// Without this rule, we might have an index which exists and contains
+	// a column which does not exist. This would lead to panics inside the
+	// optimizer and an invalid table descriptor.
+	registerDepRule("indexes containing columns reach absent before column",
+		scgraph.Precedence,
+		"index", "column",
+		func(from, to nodeVars) rel.Clauses {
+			ct := rel.Var("column-type")
+			return rel.Clauses{
+				from.el.Type((*scpb.PrimaryIndex)(nil), (*scpb.SecondaryIndex)(nil)),
+				to.el.Type((*scpb.Column)(nil)),
+				indexContainsColumn(
+					from.el, to.el, "index-column", "table-id", "column-id", "index-id",
+				),
+				ct.Type((*scpb.ColumnType)(nil)),
+				joinOnColumnID(to.el, ct, "table-id", "column-id"),
+				targetStatusEq(from.target, to.target, scpb.ToAbsent),
+				currentStatusEq(from.node, to.node, scpb.Status_ABSENT),
+				rel.Filter("columnTypeIsNotBeingDropped", ct)(func(
+					ct *scpb.ColumnType,
+				) bool {
+					return !ct.IsRelationBeingDropped
+				}),
```

Comment:
1. I don't quite understand the purpose of `ct` here, which is a ColumnType element?
2. Why do we need this clause that uses the `isRelationBeingDropped` field, which I thought was deemed "hacky"?

Reply:
I think this is needed to gracefully handle `DROP TABLE`. IIRC we wanted to replace this filter (and thus remove this field) by using better rule constructs, something along the lines of "this rule only applies on this column if the parent table isn't present in the rel-database with an ABSENT target".